File: benchmark_qc_like_phase_guided.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

#ifdef _OPENMP
#include <omp.h>
#endif

using Clock = std::chrono::high_resolution_clock;

static inline uint8_t keystream_byte(uint64_t key, uint64_t idx) {
    uint64_t x = (key * 0x9E3779B185EBCA87ULL + (idx + 1ULL) * 0xC2B2AE3D27D4EB4FULL);
    x ^= (x >> 33);
    x *= 0xFF51AFD7ED558CCDULL;
    x ^= (x >> 33);
    return static_cast<uint8_t>(x & 0xFFULL);
}

static std::vector<uint8_t> encrypt(const std::vector<uint8_t>& pt, uint64_t key) {
    std::vector<uint8_t> ct(pt.size());
    for (size_t i = 0; i < pt.size(); ++i) {
        ct[i] = static_cast<uint8_t>(pt[i] ^ keystream_byte(key, i));
    }
    return ct;
}

static bool key_matches(const std::vector<uint8_t>& ct, const std::vector<uint8_t>& known, uint64_t key) {
    for (size_t i = 0; i < known.size(); ++i) {
        if ((ct[i] ^ keystream_byte(key, i)) != known[i]) return false;
    }
    return true;
}

static std::pair<uint64_t, uint64_t> brute_force(const std::vector<uint8_t>& ct, const std::vector<uint8_t>& known, int bits) {
    const uint64_t keyspace = 1ULL << bits;
    uint64_t attempts = 0;
    for (uint64_t k = 0; k < keyspace; ++k) {
        ++attempts;
        if (key_matches(ct, known, k)) return {k, attempts};
    }
    return {UINT64_MAX, attempts};
}

static double phase_guided_score(uint64_t key, const std::vector<uint8_t>& ct, const std::vector<uint8_t>& known, int precheck_len) {
    double phi_a = (static_cast<double>(key & 0xFFULL) / 255.0) * (2.0 * M_PI);
    double phi_b = (static_cast<double>((key >> 8) & 0xFFULL) / 255.0) * (2.0 * M_PI);

    constexpr double coupling = 1.9;
    constexpr double leak = 0.02;
    constexpr double dt = 0.015;
    constexpr double anchor_hz = 200.0;
    constexpr double anchor_amp = 0.4;

    double err_acc = 0.0;
    const int n = std::min<int>(precheck_len, static_cast<int>(known.size()));
    for (int i = 0; i < n; ++i) {
        uint8_t pred = static_cast<uint8_t>(ct[i] ^ keystream_byte(key, static_cast<uint64_t>(i)));
        err_acc += std::abs(static_cast<int>(pred) - static_cast<int>(known[i])) / 255.0;
    }
    const double err = err_acc / std::max(1, n);

    double t = 0.0;
    for (int i = 0; i < 8; ++i) {
        const double d = std::sin(phi_b - phi_a);
        const double drive = (1.0 - err) * 1.5;
        const double dphi_a = coupling * d + anchor_amp * std::sin(2.0 * M_PI * anchor_hz * t - phi_a) - leak * phi_a;
        const double dphi_b = -coupling * d + drive * std::cos(phi_b) + anchor_amp * std::sin(2.0 * M_PI * anchor_hz * t - phi_b) - leak * phi_b;
        phi_a = std::fmod(phi_a + dt * dphi_a, 2.0 * M_PI);
        phi_b = std::fmod(phi_b + dt * dphi_b, 2.0 * M_PI);
        if (phi_a < 0) phi_a += 2.0 * M_PI;
        if (phi_b < 0) phi_b += 2.0 * M_PI;
        t += dt;
    }

    const double coherence = std::abs(std::cos(phi_a - phi_b));
    return err + 0.15 * coherence;
}

static std::pair<uint64_t, uint64_t> phase_guided_search(const std::vector<uint8_t>& ct, const std::vector<uint8_t>& known, int bits, int precheck_len) {
    const uint64_t keyspace = 1ULL << bits;
    std::vector<std::pair<double, uint64_t>> scored(keyspace);

#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
    for (int64_t k = 0; k < static_cast<int64_t>(keyspace); ++k) {
        scored[static_cast<size_t>(k)] = {phase_guided_score(static_cast<uint64_t>(k), ct, known, precheck_len), static_cast<uint64_t>(k)};
    }

    std::sort(scored.begin(), scored.end(), [](const auto& a, const auto& b) { return a.first < b.first; });

    uint64_t attempts = 0;
    for (const auto& sk : scored) {
        ++attempts;
        if (key_matches(ct, known, sk.second)) return {sk.second, attempts};
    }
    return {UINT64_MAX, attempts};
}

static std::vector<uint64_t> make_fixed_keys(uint64_t keyspace, int reps) {
    std::vector<double> qs;
    if (reps <= 1) qs = {0.5};
    else if (reps == 2) qs = {0.30, 0.70};
    else if (reps == 3) qs = {0.20, 0.50, 0.80};
    else {
        qs.resize(static_cast<size_t>(reps));
        for (int i = 0; i < reps; ++i) qs[static_cast<size_t>(i)] = static_cast<double>(i + 1) / static_cast<double>(reps + 1);
    }
    std::vector<uint64_t> keys;
    keys.reserve(qs.size());
    for (double q : qs) {
        uint64_t k = static_cast<uint64_t>(q * static_cast<double>(keyspace));
        if (k >= keyspace) k = keyspace - 1;
        keys.push_back(k);
    }
    return keys;
}

static int reps_for_bits(int bits) {
    if (bits <= 10) return 6;
    if (bits <= 12) return 4;
    if (bits <= 14) return 3;
    return 2;
}

struct Row {
    int bits = 0;
    uint64_t keyspace = 0;
    int reps = 0;
    std::vector<uint64_t> keys_tested;
    double bf_attempts = 0.0;
    double pg_attempts = 0.0;
    double attempt_gain = 0.0;
    double bf_time = 0.0;
    double pg_time = 0.0;
    double time_gain = 0.0;
    double grover = 0.0;
};

static std::string join_keys(const std::vector<uint64_t>& v) {
    std::ostringstream oss;
    oss << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i) oss << ", ";
        oss << v[i];
    }
    oss << "]";
    return oss.str();
}

int main(int argc, char** argv) {
    std::vector<int> bits = {8, 10, 12, 14, 16};
    int known_len = 10;
    int precheck_len = 4;
    std::string out_json = "reports/qc_like_phase_guided_benchmark_cpp.json";
    std::string out_md = "reports/qc_like_phase_guided_benchmark_cpp.md";

    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--known-len" && i + 1 < argc) known_len = std::stoi(argv[++i]);
        else if (a == "--precheck-len" && i + 1 < argc) precheck_len = std::stoi(argv[++i]);
        else if (a == "--out-json" && i + 1 < argc) out_json = argv[++i];
        else if (a == "--out-md" && i + 1 < argc) out_md = argv[++i];
    }

    const std::string base = "REZON_PHASE_GUIDED_DECRYPT_BENCH_";
    std::vector<uint8_t> pt;
    while (pt.size() < 96) pt.insert(pt.end(), base.begin(), base.end());
    pt.resize(96);
    std::vector<uint8_t> known(pt.begin(), pt.begin() + known_len);

    std::vector<Row> rows;
    for (int b : bits) {
        Row r;
        r.bits = b;
        r.keyspace = 1ULL << b;
        r.reps = reps_for_bits(b);
        r.keys_tested = make_fixed_keys(r.keyspace, r.reps);
        std::vector<double> bf_a, pg_a, bf_t, pg_t;
        for (uint64_t key : r.keys_tested) {
            auto ct = encrypt(pt, key);

            auto t0 = Clock::now();
            auto [bf_key, bf_attempts] = brute_force(ct, known, b);
            auto t1 = Clock::now();
            auto [pg_key, pg_attempts] = phase_guided_search(ct, known, b, precheck_len);
            auto t2 = Clock::now();

            if (bf_key != key || pg_key != key) {
                std::cerr << "Key recovery mismatch at bits=" << b << " key=" << key << "\n";
                return 2;
            }

            bf_a.push_back(static_cast<double>(bf_attempts));
            pg_a.push_back(static_cast<double>(pg_attempts));
            bf_t.push_back(std::chrono::duration<double>(t1 - t0).count());
            pg_t.push_back(std::chrono::duration<double>(t2 - t1).count());
        }
        auto mean = [](const std::vector<double>& x) { return std::accumulate(x.begin(), x.end(), 0.0) / std::max<size_t>(1, x.size()); };
        r.bf_attempts = mean(bf_a);
        r.pg_attempts = mean(pg_a);
        r.bf_time = mean(bf_t);
        r.pg_time = mean(pg_t);
        r.attempt_gain = r.bf_attempts / std::max(1e-12, r.pg_attempts);
        r.time_gain = r.bf_time / std::max(1e-12, r.pg_time);
        r.grover = std::sqrt(static_cast<double>(r.keyspace));
        rows.push_back(r);
    }

    {
        std::ofstream jf(out_json);
        jf << "{\n";
        jf << "  \"config\": {\n";
        jf << "    \"bits\": [8, 10, 12, 14, 16],\n";
        jf << "    \"known_len\": " << known_len << ",\n";
        jf << "    \"precheck_len\": " << precheck_len << ",\n";
#ifdef _OPENMP
        jf << "    \"openmp\": true,\n";
        jf << "    \"threads\": " << omp_get_max_threads() << "\n";
#else
        jf << "    \"openmp\": false\n";
#endif
        jf << "  },\n";
        jf << "  \"rows\": [\n";
        for (size_t i = 0; i < rows.size(); ++i) {
            const auto& r = rows[i];
            jf << "    {\n";
            jf << "      \"bits\": " << r.bits << ",\n";
            jf << "      \"keyspace\": " << r.keyspace << ",\n";
            jf << "      \"reps\": " << r.reps << ",\n";
            jf << "      \"keys_tested\": " << join_keys(r.keys_tested) << ",\n";
            jf << "      \"bruteforce_avg_attempts\": " << std::fixed << std::setprecision(6) << r.bf_attempts << ",\n";
            jf << "      \"phase_guided_avg_attempts\": " << r.pg_attempts << ",\n";
            jf << "      \"attempt_gain_x\": " << r.attempt_gain << ",\n";
            jf << "      \"bruteforce_avg_time_s\": " << r.bf_time << ",\n";
            jf << "      \"phase_guided_avg_time_s\": " << r.pg_time << ",\n";
            jf << "      \"time_gain_x\": " << r.time_gain << ",\n";
            jf << "      \"grover_ideal_queries\": " << r.grover << "\n";
            jf << "    }" << (i + 1 < rows.size() ? "," : "") << "\n";
        }
        jf << "  ]\n";
        jf << "}\n";
    }

    {
        std::ofstream mf(out_md);
        mf << "# Phase-Guided Decryption Benchmark (C++)\n\n";
        mf << "| bits | keyspace | bf avg attempts | phase-guided avg attempts | attempts gain | bf avg time (s) | pg avg time (s) | time gain | Grover ref |\n";
        mf << "|---:|---:|---:|---:|---:|---:|---:|---:|---:|\n";
        for (const auto& r : rows) {
            mf << "| " << r.bits << " | " << r.keyspace << " | "
               << std::fixed << std::setprecision(2) << r.bf_attempts << " | " << r.pg_attempts << " | "
               << std::setprecision(3) << r.attempt_gain << " | "
               << std::setprecision(6) << r.bf_time << " | " << r.pg_time << " | "
               << std::setprecision(3) << r.time_gain << " | "
               << std::setprecision(2) << r.grover << " |\n";
        }
    }

    std::cout << "saved: " << out_json << "\n";
    std::cout << "saved: " << out_md << "\n";
    return 0;
}

File: legacy/benchmark_qc_like_phase_guided.py
#!/usr/bin/env python3
import argparse
import json
import math
import time
from pathlib import Path

import numpy as np


def keystream_byte(key: int, idx: int) -> int:
    x = (key * 0x9E3779B185EBCA87 + (idx + 1) * 0xC2B2AE3D27D4EB4F) & ((1 << 64) - 1)
    x ^= x >> 33
    x = (x * 0xFF51AFD7ED558CCD) & ((1 << 64) - 1)
    x ^= x >> 33
    return x & 0xFF


def encrypt(pt: bytes, key: int) -> bytes:
    return bytes([b ^ keystream_byte(key, i) for i, b in enumerate(pt)])


def key_matches(ct: bytes, known_pt: bytes, key: int) -> bool:
    for i, b in enumerate(known_pt):
        if (ct[i] ^ keystream_byte(key, i)) != b:
            return False
    return True


def brute_force(ct: bytes, known_pt: bytes, bits: int):
    attempts = 0
    for k in range(1 << bits):
        attempts += 1
        if key_matches(ct, known_pt, k):
            return k, attempts
    return None, attempts


def phase_guided_score(key: int, ct: bytes, known_pt: bytes, precheck_len: int) -> float:
    # RC-inspired analog score:
    # 1) inject residual error signal from short known prefix
    # 2) evolve simple coupled phase state
    # 3) use final coherence error as rank score (lower is better)
    phi_a = (key & 0xFF) / 255.0 * 2.0 * np.pi
    phi_b = ((key >> 8) & 0xFF) / 255.0 * 2.0 * np.pi
    coupling = 1.9
    leak = 0.02
    dt = 0.015
    anchor_hz = 200.0
    anchor_amp = 0.4

    err_acc = 0.0
    n = min(precheck_len, len(known_pt))
    for i in range(n):
        pred = ct[i] ^ keystream_byte(key, i)
        err_acc += abs(pred - known_pt[i]) / 255.0
    err = err_acc / max(1, n)

    # 8 micro-steps of phase evolution from error drive
    t = 0.0
    for _ in range(8):
        d = np.sin(phi_b - phi_a)
        drive = (1.0 - err) * 1.5
        dphi_a = coupling * d + anchor_amp * np.sin(2.0 * np.pi * anchor_hz * t - phi_a) - leak * phi_a
        dphi_b = -coupling * d + drive * np.cos(phi_b) + anchor_amp * np.sin(2.0 * np.pi * anchor_hz * t - phi_b) - leak * phi_b
        phi_a = (phi_a + dt * dphi_a) % (2.0 * np.pi)
        phi_b = (phi_b + dt * dphi_b) % (2.0 * np.pi)
        t += dt

    # Favor low error + coherent anti-phase structure.
    coherence = abs(np.cos(phi_a - phi_b))
    return float(err + 0.15 * coherence)


def phase_guided_search(ct: bytes, known_pt: bytes, bits: int, precheck_len: int):
    keyspace = 1 << bits
    keys = np.arange(keyspace, dtype=np.int64)
    scores = np.empty(keyspace, dtype=np.float64)
    for i, k in enumerate(keys):
        scores[i] = phase_guided_score(int(k), ct, known_pt, precheck_len)
    order = keys[np.argsort(scores)]

    attempts = 0
    for k in order:
        attempts += 1
        if key_matches(ct, known_pt, int(k)):
            return int(k), attempts
    return None, attempts


def make_fixed_keys(keyspace: int, reps: int):
    if reps <= 1:
        qs = [0.5]
    elif reps == 2:
        qs = [0.30, 0.70]
    elif reps == 3:
        qs = [0.20, 0.50, 0.80]
    else:
        qs = [(i + 1) / (reps + 1) for i in range(reps)]
    return [min(keyspace - 1, max(0, int(q * keyspace))) for q in qs]


def reps_for_bits(bits: int) -> int:
    if bits <= 10:
        return 6
    if bits <= 12:
        return 4
    if bits <= 14:
        return 3
    return 2


def run(bits_list, known_len: int, precheck_len: int):
    pt = (b"REZON_PHASE_GUIDED_DECRYPT_BENCH_" * 3)[:96]
    known = pt[:known_len]
    rows = []
    for bits in bits_list:
        keyspace = 1 << bits
        reps = reps_for_bits(bits)
        keys = make_fixed_keys(keyspace, reps)

        bf_attempts, bf_time = [], []
        pg_attempts, pg_time = [], []
        for true_key in keys:
            ct = encrypt(pt, true_key)

            t0 = time.perf_counter()
            found_bf, a_bf = brute_force(ct, known, bits)
            bf_time.append(time.perf_counter() - t0)
            bf_attempts.append(a_bf)

            t1 = time.perf_counter()
            found_pg, a_pg = phase_guided_search(ct, known, bits, precheck_len=precheck_len)
            pg_time.append(time.perf_counter() - t1)
            pg_attempts.append(a_pg)

            if found_bf != true_key or found_pg != true_key:
                raise RuntimeError(f"key recovery mismatch for bits={bits}, key={true_key}")

        bf_avg = float(np.mean(bf_attempts))
        pg_avg = float(np.mean(pg_attempts))
        rows.append(
            {
                "bits": bits,
                "keyspace": keyspace,
                "reps": reps,
                "keys_tested": keys,
                "bruteforce_avg_attempts": bf_avg,
                "phase_guided_avg_attempts": pg_avg,
                "attempt_gain_x": float(bf_avg / max(pg_avg, 1e-9)),
                "bruteforce_avg_time_s": float(np.mean(bf_time)),
                "phase_guided_avg_time_s": float(np.mean(pg_time)),
                "time_gain_x": float(np.mean(bf_time) / max(np.mean(pg_time), 1e-12)),
                "grover_ideal_queries": float(math.sqrt(keyspace)),
            }
        )
    return rows


def to_markdown(rows):
    out = []
    out.append("# Phase-Guided Decryption Benchmark")
    out.append("")
    out.append("| bits | keyspace | bf avg attempts | phase-guided avg attempts | attempts gain (bf/pg) | bf avg time (s) | pg avg time (s) | time gain (bf/pg) | Grover ref queries |")
    out.append("|---:|---:|---:|---:|---:|---:|---:|---:|---:|")
    for r in rows:
        out.append(
            f"| {r['bits']} | {r['keyspace']} | {r['bruteforce_avg_attempts']:.2f} | {r['phase_guided_avg_attempts']:.2f} | "
            f"{r['attempt_gain_x']:.3f} | {r['bruteforce_avg_time_s']:.6f} | {r['phase_guided_avg_time_s']:.6f} | "
            f"{r['time_gain_x']:.3f} | {r['grover_ideal_queries']:.2f} |"
        )
    out.append("")
    out.append("Notes:")
    out.append("- Phase-guided search reorders candidates using RC-inspired analog scoring.")
    out.append("- Gains in attempts do not guarantee runtime gains due to ranking overhead.")
    return "\n".join(out)


def main():
    ap = argparse.ArgumentParser(description="Benchmark phase-guided candidate ordering vs brute-force.")
    ap.add_argument("--bits", type=str, default="8,10,12,14,16")
    ap.add_argument("--known-len", type=int, default=10)
    ap.add_argument("--precheck-len", type=int, default=4)
    ap.add_argument("--out-json", type=str, default="reports/qc_like_phase_guided_benchmark.json")
    ap.add_argument("--out-md", type=str, default="reports/qc_like_phase_guided_benchmark.md")
    args = ap.parse_args()

    bits_list = [int(x.strip()) for x in args.bits.split(",") if x.strip()]
    rows = run(bits_list, known_len=args.known_len, precheck_len=args.precheck_len)
    payload = {
        "config": {
            "bits": bits_list,
            "known_len": int(args.known_len),
            "precheck_len": int(args.precheck_len),
        },
        "rows": rows,
    }

    out_json = Path(args.out_json)
    out_json.parent.mkdir(parents=True, exist_ok=True)
    out_json.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    out_md = Path(args.out_md)
    out_md.write_text(to_markdown(rows), encoding="utf-8")

    print(json.dumps(payload, indent=2))
    print(f"saved: {out_json}")
    print(f"saved: {out_md}")


if __name__ == "__main__":
    main()
